text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,299 @@
+---
+title: SSR Rehydration
+status: DRAFTED
+created_at: 2021-08-31
+updated_at: 2021-08-31
+pr: (leave this empty until the PR is created)
+---
+
+# SSR Rehydration
+
+## Summary
+
+Whereas server-side rendering (SSR) allows a web application to be rendered into HTML on the server, rehydration involves receiving that HTML on the client and reusing the resulting DOM during initial client-side rendering (CSR).
+
+This document proposes an approach to supporting SSR rehydration in LWC.
+
+## Problem description
+
+### Motivation
+
+The motivations for SSR are [well documented](https://medium.com/walmartglobaltech/the-benefits-of-server-side-rendering-over-client-side-rendering-5d07ff2cefe8) and can be boiled down to the following two points:
+
+- SSR allows search engines to easily index URLs that would otherwise be rendered on the client as part of a web application. This results in improved [SEO](https://developers.google.com/search/docs/beginner/seo-starter-guide), with all the associated benefits.
+
+- SSR can improve the perceived startup performance of a web application by reducing the time to [First Contentful Paint](https://web.dev/first-contentful-paint/). Optimizing perceived performance can have measurable impact on [conversion rates](https://www.cloudflare.com/learning/performance/more/website-performance-conversion-rates/), [user satisfaction](https://www.ericsson.com/en/press-releases/2016/2/streaming-delays-mentally-taxing-for-smartphone-users-ericsson-mobility-report), and [revenue](https://www.dareboost.com/en/webperf-impacts).
+
+However, rendering on the server can itself introduce performance issues, working against the original intent. Namely, if DOM that was generated from SSR HTML is not preserved during the initial CSR, this will result in potentially expensive DOM manipulation and unwanted visual artifacts.
+
+The aim of rehydration is to avoid the client-side performance issues of a naive SSR implementation, reusing original DOM wherever possible.
+
+### The Happy Path
+
+With regard to SSR, there exists a happy path (Fig 1a) where the rendering output for a given request is identical on both the server and the client; identical VDOM is generated for the entire document in both environments.
+
+Depending upon the actual implementation, exceptions to this happy path include:
+
+- **Exception A:** One or more components are never rendered on the server (Fig 1b). Instead, a placeholder is rendered on the server, and later replaced on the client during CSR/hydration.
+  - **Example:** The page contains a cart icon with a dynamic number of items in cart.
+  - **Example:** The page contains a user avatar.
+- **Exception B:** One or more components are rendered only on the server (Fig 1c). No JS is sent to the client for these components and the original SSR'd HTML remains untouched by hydration and later client-side renders.

```

Comment:
> No JS is sent to the client

If no JS is sent to the client then it's not a web component, so it's just an `UnknownHTMLElement`?

Reply:
It sounds like this is basically server-side templating which won't be convenient for SPAs but great for MPAs?

Reply:
Yes, this is conceptually comparable to React server components. There are a lot of downsides to using something like this, especially (as you point out) if you want a true SPA with page-level transitions handled by JS. In LWC, you would end up with `UnknownHTMLElements` on the page, yes.

Reply:
That's a fair question.  To me, this incongruity is a natural outcome of marrying SSR (which happens outside the browser) with LWC (which "should eventually fade away to browser standards").

However, I hear you that having custom elements in the DOM without actually defining them is odd, and possibly not the right direction to go. This is a real use-case that we should either handle or explicitly _not_ support. I'll be interested to hear what other folks have to say.

Reply:
I'm trying to think of a case where `UnknownHTMLElement` would cause problems for a static component, but nothing comes to mind. Even `delegatesFocus` apparently has [a counterpart](https://github.com/mfreed7/declarative-shadow-dom/blob/a52e70f560e30c8aac2543396b1706e21445c2db/README.md#-behavior) in declarative shadow DOM.